server/src/main/java/org/opensearch/plugins/ExtensionAwarePlugin.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugins;
+
+import org.opensearch.common.settings.Setting;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Plugin that provides extra settings for extensions
+ *
+ * @opensearch.experimental
+ */
+public interface ExtensionAwarePlugin {
```

Comment:
We have an interface on SDK for extension points [here](https://github.com/opensearch-project/opensearch-sdk-java/blob/main/src/main/java/org/opensearch/sdk/Extension.java). We should name this interface as `Extension` as well since both have covers extension points. This one is for default ones for OpenSearch and the one of SDK is for extension specific.

Reply:
Hm ... this is a plugin, so in core all plugin interfaces are suffixed with `XxxPlugin`, from the other side - this not an extension, but the extension point (so confusing) for plugin to alter existing `Extension`s ...

Reply:
There is an extension point in ActionPlugin called getRestHandlerWrapper that is seemingly tailor-made for the security plugin: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/ActionPlugin.java#L130-L153

That's an important extension point because it can only be implemented once and its use is for the security plugin to wrap REST handlers and provide authentication for REST Requests before the original handler can handle the request.

I'm not sure if there would be a use-case for another plugin to define settings for extensions that should be placed in `extensions.yml`, but if there is a use-case then this extension point or a more targeted extension-specific extension point would be how a plugin can extend the default settings.

For the security plugin the values in `extensions.yml` is not enough along (similar to `opensearch.yml`)

```
extensions:
  - name: hello-world
    uniqueId: opensearch-sdk-java-1
    hostAddress: '127.0.0.1'
    port: '4532'
    version: '1.0'
    opensearchVersion: '3.0.0'
    minimumCompatibleVersion: '3.0.0'
    # example security settings
    extension_dns:
      - CN=extension-node-01,OU=SECURITY,O=OPENSEARCH,L=BROOKLYN,ST=NEW YORK,C=US
      - CN=extension-node-02,OU=SECURITY,O=OPENSEARCH,L=BROOKLYN,ST=NEW YORK,C=US
    reserved_indices: ['.hello-world-jobs']
    reserved_roles: # roles that can be mapped to a user to allow user to use extension
       ...
```

Reply:
@reta I should have mentioned it more clearly. My bad. I was thinking about `ExtensionPlugin` but again that won't make any sense.

Reply:
@cwperks thanks for clarification

> There is an extension point in ActionPlugin called getRestHandlerWrapper that is seemingly tailor-made for the security plugin: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/ActionPlugin.java#L130-L153

It strikes me that we don't need the `ExtensionAwarePlugin` and could wire everything using ActionPlugin::getRestHandlerWrapper ?

Reply:
It's not possible to use `ActionPlugin::getRestHandlerWrapper` for this purpose of extension `extensions.yml` and adding settings from a plugin.